### PR TITLE
rbd: removed hardcoded default pool

### PR DESCRIPTION
**Pull request by dillaman**

### CODE REVIEW

File: src/tools/rbd_nbd/rbd-nbd.cc [hunk 1]
```diff
@@ -792,8 +800,11 @@ static int parse_imgpath(const std::string &imgpath)
     return -EINVAL;
   }
 
-  if (match[1].matched)
+  if (match[1].matched) {
     poolname = match[1];
+  } else {
+    poolname = "";
+  }
```
Comment (liewegas) [inline src/tools/rbd_nbd/rbd-nbd.cc]: I don't understand what this is doing

File: src/common/config_opts.h [hunk 2]
```diff
@@ -1390,6 +1390,7 @@ OPTION(rbd_mirroring_replay_delay, OPT_INT, 0) // time-delay in seconds for rbd-
  * rbd_create3()/RBD::create3() and rbd_clone2/RBD::clone2() are only
  * affected by rbd_default_order.
  */
+OPTION(rbd_default_pool, OPT_STR, "rbd") // default pool for storing images
```
Comment (trociny) [inline src/common/config_opts.h]: Placing this option here might be confusing. The comment above says "rbd_create()/RBD::create() are affected by all of these options", while actually they are not affected by "rbd_default_pool".

File: src/tools/rbd/Utils.cc [hunk 3]
```diff
@@ -867,8 +920,16 @@ int init_io_ctx(librados::Rados &rados, const std::string &pool_name,
                 librados::IoCtx *io_ctx) {
   int r = rados.ioctx_create(pool_name.c_str(), *io_ctx);
   if (r < 0) {
-    std::cerr << "rbd: error opening pool " << pool_name << ": "
-              << cpp_strerror(r) << std::endl;
+    std::string default_pool_name;
+    get_default_pool_name(&default_pool_name);
+    if (r == -ENOENT && pool_name == default_pool_name) {
+      std::cerr << "rbd: error opening default pool '" << pool_name << std::endl
```
Comment (trociny) [inline src/tools/rbd/Utils.cc]: The closing quote (') after pool_name is missing.

File: src/common/config_opts.h (hunk 2 above)
Comment (trociny) [inline src/common/config_opts.h]: Just an idea and might be done as a separate PR. We could add a validator for this option. This would make error messages have more sense when e.g. rbd_default_pool is set to empty.

File: src/tools/rbd/Utils.cc [hunk 4]
```diff
@@ -590,6 +630,19 @@ int get_pool_journal_names(const po::variables_map &vm,
   return 0;
 }
 
+int validate_pool_name(const std::string &pool_name,
+                       SpecValidation spec_validation) {
+  if (spec_validation != SPEC_VALIDATION_FULL) {
+    return 0;
+  }
+
+  boost::regex pattern("^[^@/]*?$");
```
Comment (trociny) [inline src/tools/rbd/Utils.cc]: This also matches an empty string, making it a valid pool name.

File: src/tools/rbd/Utils.cc [hunk 5]
```diff
@@ -184,22 +189,35 @@ std::string get_positional_argument(const po::variables_map &vm, size_t index) {
   return "";
 }
 
-std::string get_pool_name(const po::variables_map &vm,
-                          size_t *arg_index) {
-  std::string pool_name;
+int get_default_pool_name(std::string *pool_name) {
+  std::string default_pool_name = g_ceph_context->_conf->rbd_default_pool;
+  int r = validate_pool_name(default_pool_name, SPEC_VALIDATION_FULL);
```
Comment (trociny) [inline src/tools/rbd/Utils.cc]: If we added the config option validator we could rely on it here.

**Comment by dillaman** [comment]
@trociny update pushed

File: src/common/config_validators.cc [hunk 6]
```diff
@@ -9,6 +9,27 @@
...
 #include <boost/algorithm/string.hpp>
 #include <boost/lexical_cast.hpp>
+#include <boost/regex.hpp>
+
+int validate(md_config_t::option_rbd_default_pool_t *,
+             std::string *value, std::string *error_message) {
+  boost::regex pattern("^[^@/]+$");
+  if (!boost::regex_match (*value, pattern)) {
+    *error_message = "ignoring invalid RBD default pool";
+    *value = "rbd";
```
Comment (trociny) [inline src/common/config_validators.cc]: May be we should somehow mention int the error message that 'rbd' is used instead? Because now when I have
```
% ./bin/rbd --rbd-default-pool '/sjsj' create test2 --size 128M 
parse error setting 'rbd_default_pool' to '/sjsj' (ignoring invalid RBD default pool)
```
It is not clear that the image is actually created in rbd pool.

Comment (dillaman) [inline src/common/config_validators.cc]: @trociny tweaked the message
